src/slack.py

```diff
@@ -0,0 +1,127 @@
+import json
+import os
+import random
+import requests
+import sys
+import time
+
+
+# Constants
+
+
+bearer_token = 'Bearer ' + os.getenv('API_KEY')
+IMAGE_URL = 'https://raw.githubusercontent.com/cuappdev/bagel-bot/master/bagel-logo.png'
+SLACK_API = 'https://slack.com/api/'
+TESTING_CHANNEL_ID = 'CTF81MFH6'
+
+
+# Pull / Get Request
+
+
+def pr(endpoint, data):
+    return requests.post(
+        SLACK_API + endpoint,
+        data = json.dumps(data),
+        headers = {
+            'Authorization': bearer_token,
+            'Content-type': 'application/json'
+        }
+    ).json()
+
+
+def gr(endpoint, params=None):
+    return requests.get(
+        SLACK_API + endpoint,
+        params = params,
+        headers = {
+            'Authorization': bearer_token,
+            'Content-type': 'application/x-www-form-urlencoded'
+        }
+    ).json()
+
+
+# Profiles
+
+
+profile_cache = {}
+
+
+def profile_get(slack_id):
+    global profile_cache
+    if slack_id in profile_cache:
+        return profile_cache[slack_id]
+        
+    res = gr('users.profile.get', { 'user': slack_id })
+    profile_cache[slack_id] = res['profile']
+    return res['profile']
+
+
+def name_get(slack_id):
+    return profile_get(slack_id)['real_name']
+
+
+# Messaging
+
+
+def channels_get_all():
+    res = gr('users.conversations')
+    return res['channels']
+
+
+def channel_leave(channel_id):
+    pr('conversations.leave', {
+        'channel': channel_id
+    })
+
+
+def message_channel(message, channel_id):
+    return pr('chat.postMessage', {
+        'channel': channel_id,
+        'icon_url': IMAGE_URL,
+        'text': message,
+        'username': 'bagel',
+    })
+
+
+def message_all_channels(message):
+    res = gr('users.conversations')
+    for channel in res['channels']:
+        time.sleep(5)
+        message_channel(message, channel['id'])
```

Comment:
sleeping is fine for now, but eventually, let's switch this to use `asyncio` for sending out all these. Add a todo pls?

Reply:
I'm just going to remove this method: the sleep was here to prevent the bot from immediately spamming every channel if something went wrong.

Reply:
ahh okay